PROTOCOL.md

```diff
@@ -0,0 +1,231 @@
+stream. The request may also indicate the stream is still open for more data and
+no response is expected until data is finished. If the remote indicates the
+stream is closed, the request may be considered non-unary but without anymore
+stream data sent. In the case of remote close, the remote still expects to
+receive a response or stream data. For compatibility with non streaming clients,
+a request with empty flags indicates a unary request.
+
+#### Request Flags
+
+| Flag | Name          | Description                                      |
+|------|---------------|--------------------------------------------------|
+| 0x01 | Remote Closed | Non-unary, but no more data expected from remote |
+| 0x02 | Remote Stream | Non-unary, remote is still sending data          |
+
+### Response
+
+The response message is used to end a stream with data, an empty response, or
+an error. A response message is the only expected message after a unary request.
+A non-unary request does not require a response message if the server is sending
+back stream data. A non-unary stream may return a single response message but no
+other stream data may follow.
+
+#### Response Flags
+
+No response flags are defined at this time, flags should be empty.
+
+### Data
+
+The data message is used to send data on an already initialized stream. Either
+client or server may send data. A data message is not allowed on a unary stream.
+A data message should not be sent after indicating remote close to the peer. The
+last data message on a stream must set the remote closed flag.
+
+#### Data
+
+| Flag | Name          | Description                       |
+|------|---------------|-----------------------------------|
+| 0x01 | Remote Closed | No more data expected from remote |
+
+## Streaming
+
+All ttrpc requests use streams to transfer data. Unary streams will only have
+two messages sent per stream, a request from a client and a response from the
+server. Non-unary streams, however, may send any numbers of messages from the
+client and the server. This makes stream management more complicated than unary
+streams since both client and server need to track additional state. To keep
+this management as simple as possible, ttrpc minimizes the number of states and
+uses two flags instead of control frames. Each stream has two states while a
+stream is still alive, local closed and remote closed. Each peer considers
+local and remote from their own perspective and sets flags from the other peer's
+perspective. For example, if a client sends a data frame with the remote closed
+flag, that is indicating that the client is now local closed and the server
+will be remote closed. A unary operation does not need to send these flags
+since each received message always indicates remote closed. Once a peer is
+both local and remote closed, the stream is considered finished and may be
+cleaned up.
+
+Due to the asymmetric nature of the current protocol, a client should
```

Comment:
Maybe always enclose `local close` and `remote closed` in backticks?

Reply:
Updated to be more consistent on referring to those states